File: test/Ch6_SpecialFunctions/betai/test_betai.py
"""
    Unit tests of the betai function

    Expected values extracted from:
        [URL] https://www.encyclopediaofmath.org/index.php/Incomplete_beta-function
"""
import pytest
import numericalrecipes as nr
from os import system
from subprocess import check_output
from numpy.testing import assert_almost_equal


def test_betai_x0():
    """
        Checks betai output for x = 0
    """
    a, b, x = 0.5, 0.5, 0

    calculated_value = nr.betai(a, b, x)
    expected_value = 0

    assert_almost_equal(calculated_value, expected_value, 4)


def test_betai_x1():
    """
        Checks betai output for x = 1
    """
    a, b, x = 0.5, 0.5, 1

    calculated_value = nr.betai(a, b, x)
    expected_value = 1

    assert_almost_equal(calculated_value, expected_value, 4)


def test_betai_symmetry():
    """
        Checks relation Ix(a, b) = 1 - I1-x(b, a)
    """
    a, b, x = 1, 10, 0.15

    calculated_value = nr.betai(a, b, x) + nr.betai(b, a, 1 - x)
    expected_value = 1

    assert_almost_equal(calculated_value, expected_value, 4)


def test_betai_recurrence():
    """
        Checks relation Ix(a, a) = 1/2 * I[4x(1−x)](a, 1/2)
    """
    a, b, x = 2, 10, 0.115

    calculated_value = nr.betai(a, a, x) - \
                       (1/2) * nr.betai(a, 1 / 2, 4 * x * (1 - x))
    expected_value = 0

    assert_almost_equal(calculated_value, expected_value, 4)

File: test/Ch6_SpecialFunctions/betai/test_betai.jl
#=
    Unit tests of the betai function

    Expected values obtained from:
        [URL] https://www.encyclopediaofmath.org/index.php/Incomplete_beta-function
=#


@testset "test_betai_x0" begin
    #=
        Checks betai output for x = 0
    =#
    a, b, x = 0.5, 0.5, 0

    calculated_value = betai(a, b, x)
    expected_value = 0

    @test isapprox(calculated_value, expected_value; atol=1e-8)
end


@testset "test_betai_x1" begin
    #=
        Checks betai output for x = 1
    =#
    a, b, x = 0.5, 0.5, 1

    calculated_value = betai(a, b, x)
    expected_value = 1

    @test isapprox(calculated_value, expected_value; atol=1e-8)
end


@testset "test_betai_symmetry" begin
    #=
        Checks relation Ix(a, b) = 1 - Ix-1(b, a)
    =#
    a, b, x = 1, 10, 0.15

    calculated_value = betai(a, b, x) + betai(b, a, 1 - x)
    expected_value = 1

    @test isapprox(calculated_value, expected_value; atol=1e-8)
end


@testset "test_betai_recurrence" begin
    #=
        Checks relation Ix(a, b) = xIx(a−1, b) + (1−x)Ix(a, b−1)
    =#
    a, b, x = 1, 10, 0.15

    calculated_value = betai(a, b, x) - x * betai(a - 1, b, x) - 
                       (1 - x) * betai(a, b - 1, x)

    expected_value = 0

    @test isapprox(calculated_value, expected_value; atol=1e-8)
end
